debug.go

```diff
@@ -14,30 +17,67 @@ debug-disable-type = &(
     full-permanent-disable: 4
 )
 */
+
 const (
+	// DebugNotDisabled is asserted if any debug facility, even manufacturer
+	// hardware diagnostics, is currently enabled
 	DebugNotDisabled = iota
+
+	// DebugDisabled indicates all debug facilities are currently disabled. It
+	// may be possible to enable them in the future, and it may also be possible
+	// that they were enabled in the past after the target device/sub-system
+	// booted/started, but they are currently disabled.
 	DebugDisabled
+
+	// DebugDisabledSinceBoot indicates all debug facilities are currently
+	// disabled and have been so since the target device/sub-system
+	// booted/started.
 	DebugDisabledSinceBoot
+
+	// DebugPermanentDisable indicates all non-manufacturer facilities are
+	// permanently disabled such that no end user or developer cannot enable
+	// them. Only the manufacturer indicated in the OEMID claim can enable them.
+	// This also indicates that all debug facilities are currently disabled and
+	// have been so since boot/start.
 	DebugPermanentDisable
+
+	// DebugFullPermanentDisable indicates that all debug capabilities for the
+	// target device/sub-module are permanently disabled.
 	DebugFullPermanentDisable
 )
 
 // Debug models the debug-disable type
 type Debug uint
 
-// Set makes sure that the supplied val makes a good Debug claim
-func (d *Debug) Set(val uint) error {
-	switch val {
+// Validate makes sure that the receiver is a valid Debug claim
+func (d Debug) Validate() error {
+	switch d {
 	case DebugNotDisabled:
 	case DebugDisabled:
 	case DebugDisabledSinceBoot:
 	case DebugPermanentDisable:
 	case DebugFullPermanentDisable:
 	default:
-		return fmt.Errorf("out of range value %v for Debug type", val)
+		return fmt.Errorf("out of range value %v for Debug type", d)
 	}
 
-	*d = Debug(val)
-
 	return nil
 }
+
+// MarshalCBOR CBOR encodes a valid Debug receiver
+func (d Debug) MarshalCBOR() ([]byte, error) {
+	if err := d.Validate(); err != nil {
```

Comment:
Given that we're not validating on unmarshalling, does it make sense to validate on marshalling? If someone puts an invalid value in, and we accept it, would then not also expect to be able to get it out?

Reply:
Right, but Postel's principle is asymmetric :-)  (I'm applying that here.)

Reply:
I don't think that makes sense here. That talks about complete systems, this is a component. We're not creating the data we're putting out here. The point of splitting out bounds checking into a separate `Validate()` is to allow this to be used in systems where invalid data may already be present.

E.g. imagine this is used in  a system that reads a token in, passes it around (possibly routing it based on some of its values), and then writes it out, all without modifying it in any way. As an engineer implementing this I would expect the library I'm using to either, a) not care about the out of bounds values within leaving it to me check if I wish, or b) to fail early on reading the value in.

Reply:
I'm not sure that Postel is not applicable to emitters of single protocol elements, as in this case, as long as this strategy is applied all over to construct the whole protocol bottom-up.

I also think your example could be handled with what is here already: it suffices to decode the raw token while keeping it in memory somewhere, operate on the its internalised structure and then fetch the raw bytes from memory and spit it out.

However, I buy the point that this effectively mixes a bit of policy with a mechanism in a way that the user of the library cannot control completely.  And that I don't like.  So I am OK with untangling the two :-)